[ONNX] Refactor heavy memory usage tests

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #79641
* __->__ #79640

* Move memory heavy tests from `test_pytorch_onnx_onnxruntime.py` to
  `test_models_onnxruntime.py`. The former is run in parallel in CI,
  while the latter is not. A change is that the moved tests are now
  only covered in default opset export.
* Refactor and create base class for tests that export model to ONNX
  and verify with ONNX Runtime. The new base class are parameterized
  with `opset_version` and `is_script`. Further work can be done to
  refactor existing test classes in `test_pytorch_onnx_onnxruntime.py`.
  See #75630
* Reduce unnecessarily large tensor size in
  `test_pytorch_onnx_onnxruntime.py` to further reduce memory usage
  and test time.

After this PR, the running time for `test_pytorch_onnx_onnxruntime.py`
is reduced from `1338.82s (0:22:18)` to `225.07s (0:03:45)`,
benchmarked on 10900x with `-n 10`.

Fixes #79179

BowenBao committed:
[ONNX] Refactor heavy memory usage tests

* Move memory heavy tests from `test_pytorch_onnx_onnxruntime.py` to
  `test_models_onnxruntime.py`. The former is run in parallel in CI,
  while the latter is not. A change is that the moved tests are now
  only covered in default opset export.
* Refactor and create base class for tests that export model to ONNX
  and verify with ONNX Runtime. The new base class are parameterized
  with `opset_version` and `is_script`. Further work can be done to
  refactor existing test classes in `test_pytorch_onnx_onnxruntime.py`.
  See #75630
* Reduce unnecessarily large tensor size in
  `test_pytorch_onnx_onnxruntime.py` to further reduce memory usage
  and test time.

After this PR, the running time for `test_pytorch_onnx_onnxruntime.py`
is reduced from `1338.82s (0:22:18)` to `225.07s (0:03:45)`,
benchmarked on 10900x with `-n 10`.

Fixes #79179

[ghstack-poisoned]

justinchuby reviewed (approved):
🏃 ⚡

garymm commented on scripts/onnx/test.sh:
```diff
@@ -30,6 +30,7 @@ fi
 pip install pytest scipy hypothesis # these may not be necessary
 pip install pytest-cov # installing since `coverage run -m pytest ..` doesn't work
 pip install -e tools/coverage_plugins_package # allows coverage to run w/o failing due to a missing plug-in
+pip install parameterized # parameterizing test class
```
Might be faster if we combine the pip install commands

garymm commented on scripts/onnx/test.sh:
```diff
@@ -59,36 +60,29 @@ if [[ "${SHARD_NUMBER}" == "1" ]]; then
-    "$top_dir/test/onnx/test_pytorch_onnx_caffe2.py" \
-    "$top_dir/test/onnx/test_pytorch_onnx_caffe2_quantized.py"
+    "$top_dir/test/onnx/test_utility_funs.py" \
+    "$top_dir/test/onnx/test_models_onnxruntime.py" "-k" "not TestModelsONNXRuntime"
```
This `-k` thing is a bit confusing. What's left in this file besides `TestModelsONNXRuntime`? Why not move it to another file?

BowenBao replied:
There are two other test classes running model export test with ort defined from `test_model.py`. `test_model.py` tests with caffe2. I'm opening #79802 to track this.

garymm commented on test/onnx/test_models_onnxruntime.py:
```diff
@@ -41,5 +63,357 @@ def exportTest(self, model, inputs, rtol=1e-2, atol=1e-7, opset_versions=None):
+
+@parameterized.parameterized_class(
+    ("opset_version", "is_script"),
+    itertools.product([_constants.onnx_default_opset], [True, False]),
```
Why are we parameterizing with only one opset version?
Seems unnecessary, since that's the default for the base class anyways.

BowenBao replied:
Good catch, getting rid of it. Previously more versions were included, but I think it's unnecessary.

garymm commented on test/onnx/test_models_onnxruntime.py:
```diff
@@ -41,5 +63,357 @@ def exportTest(self, model, inputs, rtol=1e-2, atol=1e-7, opset_versions=None):
+            additional_test_inputs=[(images, features), (images2, test_features)],
+        )
+
+    @skipScriptTest()  # TODO: https://msdata.visualstudio.com/Vienna/_workitems/edit/1253950
```
update to https://github.com/pytorch/pytorch/issues/75625

garymm commented on test/onnx/test_pytorch_common.py:
```diff
@@ -51,12 +51,12 @@ def wrapper(*args, **kwargs):
         @functools.wraps(func)
-        def wrapper(self, *args, **kwargs):
-            if self.opset_version < min_opset_version:
+        def wrapper(test_class, *args, **kwargs):
```
Why the name change?
Is the first argument really a class? I thought it was an instance of the TestCase, in which case the old name seems better?

justinchuby replied:
I proposed the change initially. I agree after giving it another thought

BowenBao replied:
some convention uses `cls` too. `self` is clearer in a sense this is instance method.

justinchuby replied:
`test_case`? Since type(self) == TestCase

garymm reviewed (approved):
Thanks!

BowenBao committed:
Update on "[ONNX] Refactor heavy memory usage tests"


* Move memory heavy tests from `test_pytorch_onnx_onnxruntime.py` to
  `test_models_onnxruntime.py`. The former is run in parallel in CI,
  while the latter is not. A change is that the moved tests are now
  only covered in default opset export.
* Refactor and create base class for tests that export model to ONNX
  and verify with ONNX Runtime. The new base class are parameterized
  with `opset_version` and `is_script`. Further work can be done to
  refactor existing test classes in `test_pytorch_onnx_onnxruntime.py`.
  See #75630
* Reduce unnecessarily large tensor size in
  `test_pytorch_onnx_onnxruntime.py` to further reduce memory usage
  and test time.

After this PR, the running time for `test_pytorch_onnx_onnxruntime.py`
is reduced from `1338.82s (0:22:18)` to `225.07s (0:03:45)`,
benchmarked on 10900x with `-n 10`.

Fixes #79179

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Refactor heavy memory usage tests"


* Move memory heavy tests from `test_pytorch_onnx_onnxruntime.py` to
  `test_models_onnxruntime.py`. The former is run in parallel in CI,
  while the latter is not. A change is that the moved tests are now
  only covered in default opset export.
* Refactor and create base class for tests that export model to ONNX
  and verify with ONNX Runtime. The new base class are parameterized
  with `opset_version` and `is_script`. Further work can be done to
  refactor existing test classes in `test_pytorch_onnx_onnxruntime.py`.
  See #75630
* Reduce unnecessarily large tensor size in
  `test_pytorch_onnx_onnxruntime.py` to further reduce memory usage
  and test time.

After this PR, the running time for `test_pytorch_onnx_onnxruntime.py`
is reduced from `1338.82s (0:22:18)` to `225.07s (0:03:45)`,
benchmarked on 10900x with `-n 10`.

Fixes #79179

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g

BowenBao committed:
Update on "[ONNX] Refactor heavy memory usage tests"


* Move memory heavy tests from `test_pytorch_onnx_onnxruntime.py` to
  `test_models_onnxruntime.py`. The former is run in parallel in CI,
  while the latter is not. A change is that the moved tests are now
  only covered in default opset export.
* Refactor and create base class for tests that export model to ONNX
  and verify with ONNX Runtime. The new base class are parameterized
  with `opset_version` and `is_script`. Further work can be done to
  refactor existing test classes in `test_pytorch_onnx_onnxruntime.py`.
  See #75630
* Reduce unnecessarily large tensor size in
  `test_pytorch_onnx_onnxruntime.py` to further reduce memory usage
  and test time.

After this PR, the running time for `test_pytorch_onnx_onnxruntime.py`
is reduced from `1338.82s (0:22:18)` to `225.07s (0:03:45)`,
benchmarked on 10900x with `-n 10`.

Fixes #79179

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g
